src/citrine/ara/variables.py

```diff
@@ -101,3 +110,232 @@ def __init__(self, *,
+
+
+class AttributeByTemplateAndObjectTemplate(Serializable['AttributeByTemplateAndObjectTemplate'], Variable):
+    """Attribute marked by an attribute template and an object template.
+
+    TODO: clarify documentation of use case here.
+
+    Parameters
+    ---------
+    name: str
+        a short human-readable name to use when referencing the variable
+    headers: list[str]
+        sequence of column headers
+    attributeTemplate: LinkByUID
+        attribute template that identifies the attribute to assign to the variable
+    objectTemplate: LinkByUid
+        template that identifies the associated object
+    """
+
+    name = properties.String('name')
+    headers = properties.List(properties.String, 'headers')
+    attributeTemplate = properties.Object(LinkByUID, 'attributeTemplate')
+    objectTemplate = properties.Object(LinkByUID, 'objectTemplate')
+    typ = properties.String('type', default="attribute_by_object", deserializable=False)
+
+    def _attrs(self) -> List[str]:
+        return ["name", "headers", "attributeTemplate", "objectTemplate", "typ"]
+
+    def __init__(self, *,
+                 name: str,
+                 headers: List[str],
+                 attributeTemplate: LinkByUID,
+                 objectTemplate: LinkByUID):
+        self.name = name
+        self.headers = headers
+        self.attributeTemplate = attributeTemplate
+        self.objectTemplate = objectTemplate
+
+
+class IngredientIdentifierByProcessTemplateAndName(Serializable['IngredientIdentifierByProcessAndName'], Variable):
+    """Ingredient identifier associated with a process template and a name.
+
+    Parameters
+    ---------
+    name: str
+        a short human-readable name to use when referencing the variable
+    headers: list[str]
+        sequence of column headers
+    processTemplate: LinkByUID
+        process template associated with this ingredient identifier
+    ingredientName: str
+        name of ingredient
+    scope: str
+    """
+
+    name = properties.String('name')
+    headers = properties.List(properties.String, 'headers')
+    processTemplate = properties.Object(LinkByUID, 'processTemplate')
+    ingredientName = properties.String('ingredientName')
+    scope = properties.String('scope')
+    typ = properties.String('type', default="ing_id_by_process_and_name", deserializable=False)
+
+    def _attrs(self) -> List[str]:
+        return ["name", "headers", "processTemplate", "ingredientName", "scope", "typ"]
+
+    def __init__(self, *,
+                 name: str,
+                 headers: List[str],
+                 processTemplate: LinkByUID,
+                 ingredientName: str,
+                 scope: str):
+        self.name = name
+        self.headers = headers
+        self.processTemplate = processTemplate
+        self.ingredientName = ingredientName
+        self.scope = scope
+
+
+class IngredientLabelByProcessAndName(Serializable['IngredientLabelByProcessAndName'], Variable):
```

Comment:
This name makes me think that it will return the label of the ingredient with a given name and process, not a boolean indicating whether or not the ingredient has a specified label.

Reply:
Agreed. I had the same confusion. How about `IngredientHasLabelByProcessAndName`? I think that change needs to be coordinated on the backend.

Reply:
Works for me.  That backend change is easy to make (and I've been expecting to massage the names as we go).